modular_skyrat/modules/delam_emergency_stop/code/scram.dm

```diff
@@ -0,0 +1,435 @@
+	button_stage = BUTTON_ARMED
+	update_appearance()
+	radio.talk_into(src, "SUPERMATTER EMERGENCY STOP BUTTON ARMED!", RADIO_CHANNEL_ENGINEERING)
+	visible_message(span_danger("[user] swings open the plastic cover of the [src]!"))
+
+	// Let the admins know someone's fucked up
+	message_admins("[ADMIN_LOOKUPFLW(user)] just opened the cover of the [src].")
+	investigate_log("[key_name(user)] opened the cover of the [src].", INVESTIGATE_ATMOS)
+
+	confirm_action()
+
+/// Confirms with the user that they really want to push the red button. Do it, you won't!
+/obj/machinery/button/delam_scram/proc/confirm_action(mob/user, list/modifiers)
+	if(tgui_alert(usr, "You really sure that you want to push this?", "It looked scarier on HBO.", list("No", "Yes")) != "Yes")
+		button_stage = BUTTON_AWAKE
+		visible_message(span_danger("[user] slowly closes the plastic cover of the [src]!"))
+		update_appearance()
+		return
+
+	// Make scary sound and flashing light
+	playsound(src, 'sound/machines/high_tech_confirm.ogg', 50, FALSE, 15, ignore_walls = TRUE, use_reverb = TRUE)
+	button_stage = BUTTON_PUSHED
+	visible_message(span_danger("[user] smashes the [src] with their hand!"))
+	message_admins("[ADMIN_LOOKUPFLW(user)] pushed the [src]!")
+	investigate_log("[key_name(user)] pushed the [src]!", INVESTIGATE_ATMOS)
+	flick_overlay_view("[base_icon_state]-overlay-active", 20 SECONDS)
+
+	// No going back now!
+	SEND_GLOBAL_SIGNAL(COMSIG_MAIN_SM_DELAMINATING, MANUAL_INTERVENTION)
+
+	// Temporarily let anyone escape the engine room before it becomes spicy
+	for(var/obj/machinery/door/airlock/escape_route in range(7, src))
+		if(istype(escape_route, /obj/machinery/door/airlock/command))
+			continue
+
+		INVOKE_ASYNC(escape_route, TYPE_PROC_REF(/obj/machinery/door/airlock, temp_emergency_exit), 45 SECONDS)
+
+/// When the button is pushed but it's too late to save you!
+/obj/machinery/button/delam_scram/proc/burn_out()
+	if(!(machine_stat & BROKEN))
+		set_machine_stat(machine_stat | BROKEN)
+		update_appearance()
+
+/obj/machinery/button/delam_scram/update_icon_state()
+	. = ..()
+	icon_state = "[base_icon_state][skin]"
+	if(button_stage == BUTTON_ARMED)
+		icon_state += "-armed"
+	else if(button_stage == BUTTON_PUSHED)
+		icon_state += "-armed"
+	else if(machine_stat & (NOPOWER|BROKEN))
+		icon_state += "-nopower"
+
+/obj/machinery/power/supermatter_crystal/Destroy() // I wish I could set the sign to negatives if you manage to still screw it up
+	if(is_main_engine && GLOB.main_supermatter_engine == src)
+		SSpersistence.delam_highscore = SSpersistence.rounds_since_engine_exploded
+		SSpersistence.rounds_since_engine_exploded = MISTAKES_WERE_MADE
+		for (var/obj/machinery/incident_display/sign as anything in GLOB.map_delamination_counters)
+			sign.update_delam_count(MISTAKES_WERE_MADE)
+	return ..()
+
+/obj/machinery/power/emitter/LateInitialize(mapload)
+	. = ..()
+	RegisterSignal(SSdcs, COMSIG_MAIN_SM_DELAMINATING, PROC_REF(emergency_stop))
```

Comment:
Does this only work on the emitters that are in the SM chamber, or is that valid for every emitter that's currently built? I could see this potentially being a bit weird if we have emitters doing other things, suddenly shutting off because of that button.

Reply:
It used to check area, but I was told in a review comment here that it's really expensive to do an area check on the machines list.

GLOB.machines is gone now though, so probably can be put back in.